sklearn/neighbors/tests/test_neighbors.py

```diff
@@ -163,6 +163,72 @@ def test_precomputed(random_state=42):
         assert_array_almost_equal(pred_X, pred_D)
 
 
+def test_precomputed_dense():
+    def make_train_test(X_train, X_test):
+        return (metrics.pairwise_distances(X_train),
+                metrics.pairwise_distances(X_test, X_train))
+
+    check_precomputed(make_train_test)
+
+
+@pytest.mark.parametrize('fmt', ['csr', 'lil'])
+def test_precomputed_sparse_implicit_diagonal(fmt):
+    def make_train_test(X_train, X_test):
+        nn = neighbors.NearestNeighbors(n_neighbors=3).fit(X_train)
+        return (nn.kneighbors_graph(None, mode='distance').asformat(fmt),
+                nn.kneighbors_graph(X_test, mode='distance').asformat(fmt))
+
+    check_precomputed(make_train_test, implicit_diagonal=True)
+
+
+def test_precomputed_sparse_explicit_diagonal():
+    def make_train_test(X_train, X_test):
+        nn = neighbors.NearestNeighbors(n_neighbors=3).fit(X_train)
+        return (nn.kneighbors_graph(X_train.copy(), mode='distance'),
+                nn.kneighbors_graph(X_test, mode='distance'))
+
+    check_precomputed(make_train_test, implicit_diagonal=False)
+
+
+def test_is_sorted_by_data():
+    # Test the helper function _is_sorted_by_data
```

Comment:
You need to also test what happens at row endings, i.e. that data can reduce, stay the same or increase across row boundaries.

Also check what happens with ties (or just all-equal data)

Reply:
Was this addresses?